crates/aquascope/src/analysis/stepper/stepper_new.rs

```diff
@@ -0,0 +1,1296 @@
+//! Because of this invariant, the stepper uses a strategy to "slice" the MIR into segments, such that
+//! these segments always form a total cover.
+//!
+//! ### Data Structures and Invariants
+//!
+//! The key data structures involvled are the [`MirSegment`] and [`SegmentTree`]. To summarize their
+//! invariants, a `MirSegment` is a segment of the control-flow graph such that the entry dominates
+//! all locations until the exit. The corresponding exit should also post-dominate all locations.
+//! Put simply, if control-flow enters the `from` location, it will leave via the `to` location.
+//!
+//! The [`SegmentTree`] (not to be confused with a [segment tree](https://en.wikipedia.org/wiki/Segment_tree))
+//! holds all of the [`MirSegment`]s required to form a total cover.
+//!
+//! ### Slicing
+//!
+//! The core operation performed on the [`SegmentTree`] is taking a *slice*. There are two kinds of
+//! slices:
+//!
+//! 1. linear slices, those that **do not** span several branches of control-flow.
+//! 2. control-flow slices, those that **do** span several branches of control-flow.
+//!
+//! These two slices exist to maintain the invariants of the [`MirSegment`] and [`SegmentTree`].
+//! Fundamentally, these slices work on different *shapes* of the underlying graph. A *linear slice*
+//! would slice a portion of the graph which forms a line.
+//!
+//! Example:
+//! ```text
+//! before slice:
+//!
+//!             slice point
+//!               |
+//! [segment 1]   |
+//! ⬤ [l1] ----> ⬤ [l2] ----> ⬤ [l3]
+//!               |
+//!               |
+//!
+//!
+//! after slice:
+//!
+//! [segment 1]
+//! ⬤ [l1] ----> ⬤ [l2]
+//!
+//! [segment 2]
+//! ⬤ [l2] ----> ⬤ [l3]
+//!
+//! ```
+//! In the above example there exists a linear sequence of control-flow from `l1 ⟶ l2 ⟶ l3`.
+//! Depicted, is a *linear slice* of this segment at location `l2`. Linear slices *always*
+//! split a single segment, into two new segments which maintain the [`MirSegment`] invariant.
+//! These slices are used after [`Stmt`s](rustc_hir::Stmt) and the end of a [`Block` expression](rustc_hir::Block).
+//!
+//! A *control-flow* slice, then does not slice a linear sequence of locations but multiple that
+//! /span across/ branches of control flow.
+//!
+//! Example:
+//! ```text
+//! before slice:
+//!
+//!                     slice point
+//!                        |
+//!                        |
+//!                        |
+//!               ------> ⬤ [l2] -------
+//!               |        |            |
+//! [segment 1]   |        |            v
+//!        ----> ⬤ [l1]    |            ⬤ [l4] ---->
+//!               |        |            ^
+//!               |        |            |
+//!               ------> ⬤ [l3] -------
+//!                        |
+//!                        |
+//!                        |
+//!
+//!
+//! after slice:
+//!
+//! [segment 1]
+//! ⬤ [l1] ----> ⬤ [l2]
```

Comment:
Does this violate the invariant that a segment's exist post-dominates its entry? Since `l2` does not post-dominate `l1` in this example.